docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,392 @@
+    connection.query(q); // <-- add 'q' as a SQL injection sink
+  }
+
+We can recognize this using the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["mysql.Connection", "Member[query].Argument[0]", "sql-injection"]
+
+To break this down:
+
+- The first column, **"mysql.Connection"**, begins the search at any expression whose value is known to be an instance of
+  the **Connection** type from the **mysql** package. This will select the **connection** parameter above because of its type annotation.
+- **Member[query]** selects the **query** member from the connection object.
+- **Argument[0]** selects the first argument of a call to that member.
+- **sql-injection** indicates that this is considered a sink for the SQL injection query.
+
+This works for the example above because the **connection** parameter has a type annotation that matches what the model is looking for.
+
+In the next section, we'll show how to generalize the model to handle the absense of type annotations.
+
+Continued example: Dealing with untyped code
+--------------------------------------------
+
+Suppose we want the model from above to detect the sink in this snippet:
+
+.. code-block:: js
+
+  import { getConnection } from "@example/db";
+  let connection = getConnection();
+  connection.query(q); // <-- add 'q' as a SQL injection sink
+
+There is no type annotation on **connection**, and there is no indication of what **getConnection()** returns.
+Using a **typeModel** tuple we can tell our model that this function returns an instance of **mysql.Connection**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: typeModel
+      data:
+        - ["mysql.Connection", "@example/db", "Member[getConnection].ReturnValue"]
+
+To break this down:
+
+- Since we're providing type information, we add a tuple to the **typeModel** extension point.
+- The first column, **"mysql.Connection"**, names the type that we're adding a new definition for.
+- The second column, **"@example/db"**, begins the search at imports of the hypothetical NPM package **@example/db**.
+- **Member[getConnection]** selects references to the **getConnection** member from that package.
+- **ReturnValue** selects the return value from a call to that member.
+
+The new model states that the return value of **getConnection()** has type **mysql.Connection**.
+Combining this with the sink model we added earlier, the sink in the example is detected by the model.
+
+The mechanism used here is how library models work for both TypeScript and plain JavaScript.
+A good library model contains **typeModel** tuples to ensure it works even in codebases without type annotations.
+For example, the **mysql** model that is included with the CodeQL JS analysis includes this type definition (among many others):
```

Comment:
A tiny example of how you can create types, from other types, would be nice.  

Something like (bad example):  
```
["mysql.Connection", "mysql.Connection", "Member[clone].ReturnValue"]
```

Reply:
I understand the sentiment behind this, but I can't seem to find a good example to use.